spec/services/pagination_service_spec.rb

```diff
@@ -0,0 +1,689 @@
+            expect(response['links']['self']).to eq self_url
+          end
+
+          it "sets prev link to first page and next link to next offset" do
+            first_page = "#{base_url}#{url_path}?q=term&format=jsonapi&page_limit=10&page_offset=1"
+            next_url = "#{base_url}#{url_path}?q=term&format=jsonapi&page_limit=10&page_offset=14"
+            expect(response['links']['prev']).to eq first_page
+            expect(response['links']['next']).to eq next_url
+          end
+
+          it "sets first to first page and last link to third page" do
+            first_page = "#{base_url}#{url_path}?q=term&format=jsonapi&page_limit=10&page_offset=1"
+            third_page = "#{base_url}#{url_path}?q=term&format=jsonapi&page_limit=10&page_offset=31"
+            expect(response['links']['first']).to eq first_page
+            expect(response['links']['last']).to eq third_page
+          end
+        end
+      end
+
+      context 'and there are no results' do
+        let(:results) { [] }
+
+        it 'returns no results' do
+          expect(response["data"]).to match_array([])
+        end
+
+        it "sets meta['page'] stats" do
+          expect(response["meta"]["page"]["page_offset"]).to eq "1"
+          expect(response["meta"]["page"]["page_limit"]).to eq "10"
+          expect(response["meta"]["page"]["actual_page_size"]).to eq "0"
+          expect(response["meta"]["page"]["total_num_found"]).to eq "0"
+        end
+
+        it 'sets links with prev and next set to nil, last set to first' do
+          expect(response['links']["self"]).to eq first_page
+          expect(response['links']["first"]).to eq first_page
+          expect(response['links']["prev"]).to eq nil
+          expect(response['links']["next"]).to eq nil
+          expect(response['links']["last"]).to eq first_page
+        end
+
+        it 'does not include errors' do
+          expect(response.key?('errors')).to eq false
+        end
+      end
+
+      context 'and results do not include a full page' do
+        context 'and there are several results in range' do
+          let(:query_string) { 'q=term&format=jsonapi&page_offset=31&page_limit=10' }
+
+          before do
+            params[:page_offset] = "31"
+            query_params[:page_offset] = "31"
+          end
+
+          it 'returns json api response with a partial page of results starting at 11th result' do
+            expect(response["data"]).to match_array(results[30..35])
+          end
+
+          it "sets meta['page'] stats" do
+            expect(response["meta"]["page"]["page_offset"]).to eq "31"
+            expect(response["meta"]["page"]["page_limit"]).to eq "10"
+            expect(response["meta"]["page"]["actual_page_size"]).to eq "6"
+            expect(response["meta"]["page"]["total_num_found"]).to eq "36"
+          end
+
+          it "sets self to last page" do
+            expect(response['links']['self']).to eq last_page
```

Comment:
Why is this `last_page`?  I don't understand why this case is handled differently than the next one where the page_offset is 36.

Reply:
The `page_offset` (i.e., 31) in this example falls on an expected page boundary (e.g. 1, 11, 21, 31).  And 31 is the start of the `last_page` allowing the `last_page` url to be used for `response['links']['self']`.

In the next example, `page_offset` is 36 which is not on an expected page boundary.  So the defined `last_page` is not the page that was requested.  This requires `self_url` to be defined in the spec.

You didn't ask, but one of the decision points on behavior that I wrangled with is the definition of `last_page` when the `page_offset` is not on an expected page boundary.  I decided that `prev` and `next` would be calculated off the passed in `page_offset` (i.e. `page_offset -/+ page_limit`), but `last_page` would always be the last page calculated from `page_offset=1` which always puts it on an expected page boundary.

Reply:
:+1: That makes total sense and looking at this again I can't understand what I was thinking before.

Reply:
I added documentation for the class describing this process.  Let me know if it is clear enough.